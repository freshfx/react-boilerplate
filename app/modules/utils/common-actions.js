import {createAction} from '@reduxjs/toolkit'
import {serializeError} from 'serialize-error'

const resetState = initialState => state => {
  Object.keys(initialState).forEach(key => {
    state[key] = initialState[key]
  })
}

const prepareErrorAction = (payload = {}) => ({
  payload: {
    ...payload,
    ...(payload.error && {error: serializeError(payload.error)})
  }
})

const createNetworkAction = ({name, scope}) => {
  const action = createAction(`${scope}/${name}`)

  action.resolved = createAction(`${action.type}.resolved`)
  action.rejected = createAction(`${action.type}.rejected`, prepareErrorAction)

  return action
}

export {createNetworkAction, resetState}
