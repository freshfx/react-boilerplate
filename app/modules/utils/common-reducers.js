import STATUS from '../status'

const createNetworkReducers = ({networkAction, reducers}) => ({
  [networkAction.type]: (state, action) => {
    state.error = null
    state.status = STATUS.PENDING
    reducers?.[networkAction.type]?.(state, action)
  },
  [networkAction.resolved.type]: (state, action) => {
    state.status = STATUS.SUCCESS
    reducers?.[networkAction.resolved.type]?.(state, action)
  },
  [networkAction.rejected.type]: (state, action) => {
    state.error = action.payload?.error
    state.status = STATUS.FAILURE
    reducers?.[networkAction.rejected.type]?.(state, action)
  }
})

// eslint-disable-next-line import/prefer-default-export
export {createNetworkReducers}
