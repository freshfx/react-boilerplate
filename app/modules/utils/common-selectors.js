import {createSelector} from '@reduxjs/toolkit'

import {selectEntitiesGenerator} from 'modules/entities/selectors'

const selectId = (state, props) => props.id

const createSliceSelector = (name, initialState) => state =>
  state[name] || initialState

const createSliceFieldSelector = (sliceSelector, fieldName, fallback) =>
  createSelector(sliceSelector, slice => slice[fieldName] ?? fallback)

const createEntitySelectors = ({type}) => {
  const selectEntities = selectEntitiesGenerator(type)
  const selectEntity = createSelector(
    selectEntities,
    selectId,
    (entities, id) => entities[id] || {}
  )

  const createEntityFieldSelector = callback =>
    createSelector(selectEntity, callback)

  return {
    createEntityFieldSelector,
    selectEntities,
    selectEntity
  }
}

const createResultSelectors = ({initialState, scope}) => {
  const selectSlice = createSliceSelector(scope, initialState)
  return {
    selectData: createSliceFieldSelector(selectSlice, 'data'),
    selectError: createSliceFieldSelector(selectSlice, 'error'),
    selectHasNextPage: createSliceFieldSelector(selectSlice, 'hasNextPage'),
    selectStatus: createSliceFieldSelector(selectSlice, 'status'),
    selectTotal: createSliceFieldSelector(selectSlice, 'total')
  }
}

export {
  createEntitySelectors,
  createResultSelectors,
  createSliceFieldSelector,
  createSliceSelector,
  selectId
}
