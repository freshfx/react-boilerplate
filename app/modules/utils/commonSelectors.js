/* eslint-disable import/prefer-default-export */
const selectId = (state, props) => props.id

export {
  selectId
}
