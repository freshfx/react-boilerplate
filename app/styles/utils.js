import {rgba, parseToRgb} from 'polished'
import {keyframes} from '@emotion/react'

export const divide = (dividend, divisor) => {
  const [unit] = dividend.match(/[a-zA-Z%]*$/u)
  return `${parseFloat(dividend) / divisor}${unit}`
}

export const addAlpha = (color, alpha) =>
  rgba({
    ...parseToRgb(color),
    alpha
  })

const DEFAULT_OPACITY_FROM = 0
const DEFAULT_OPACITY_TO = 1
export const createOpacityAnimation = (
  from = DEFAULT_OPACITY_FROM,
  to = DEFAULT_OPACITY_TO
) => keyframes`
  from {
    opacity: ${from};
  }
  to {
    opacity: ${to};
  }
`
