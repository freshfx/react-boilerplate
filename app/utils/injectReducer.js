import React from 'react'
import hoistNonReactStatics from 'hoist-non-react-statics'
import {ReactReduxContext} from 'react-redux'

import getInjectors from './reducerInjectors'

/**
 * Dynamically injects a reducer
 *
 * @param {string} key A key of the reducer
 * @param {function} reducer A reducer that will be injected
 * @return {component} given component with reducer
 *
 */
export default ({key, reducer}) => WrappedComponent => {
  class ReducerInjector extends React.Component {
    constructor(props, context) {
      super(props, context)

      getInjectors(context.store).injectReducer(key, reducer)
    }

    static WrappedComponent = WrappedComponent

    static displayName = `withReducer(${WrappedComponent.displayName ||
      WrappedComponent.name ||
      'Component'})`

    static contextType = ReactReduxContext

    render() {
      return <WrappedComponent {...this.props} />
    }
  }

  return hoistNonReactStatics(ReducerInjector, WrappedComponent)
}
