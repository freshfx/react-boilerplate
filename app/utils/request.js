import 'whatwg-fetch'

const STATUS_CODES = {
  multipleChoices: 300,
  noContent: 204,
  ok: 200,
  resetContent: 205
}

const defaultOptions = {
  mode: 'cors'
}

/**
 * Parses the JSON returned by a network request
 *
 * @param  {object} response A response from a network request
 *
 * @return {object}          The parsed JSON from the request
 */
const parseJSON = response => response.json().catch(() => null)

/**
 * Reads the text returned by a network request
 *
 * @param  {object} response A response from a network request
 *
 * @return {string}          The USVString object (text) from the request
 */
const getText = response => response.text().catch(() => null)

/**
 * Checks if a network request came back fine, and throws an error if not
 *
 * @param  {object} response   A response from a network request
 *
 * @return {object|undefined} Returns either the response, or throws an error
 */
const checkStatus = async response => {
  if (
    response.status >= STATUS_CODES.ok &&
    response.status < STATUS_CODES.multipleChoices
  ) {
    return response
  }

  const error = new Error(response.statusText)
  error.response = response
  error.body = await parseJSON(response)
  throw error
}

/**
 * Requests a URL, returning a promise
 *
 * @param  {string} url       The URL we want to request
 * @param  {object} [options] The options we want to pass to "fetch"
 *
 * @return {string}           The USVString object (text) from the request
 */
const requestText = (url, options) =>
  fetch(url, {...defaultOptions, ...options})
    .then(checkStatus)
    .then(getText)

export {getText, requestText}

/**
 * Requests a URL, returning a promise
 *
 * @param  {string} url       The URL we want to request
 * @param  {object} [options] The options we want to pass to "fetch"
 *
 * @return {object}           The response data
 */
export default function request(url, options) {
  return fetch(url, {...defaultOptions, ...options})
    .then(checkStatus)
    .then(parseJSON)
}
