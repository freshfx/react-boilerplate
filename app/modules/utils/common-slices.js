import {createSlice} from '@reduxjs/toolkit'

import STATUS from '../status'

import {resetState} from './common-actions'

/* eslint-disable sort-keys */
const createAsyncResultsSlice = ({
  asyncThunk,
  initialState,
  scope,
  reducers,
  extraReducers
}) =>
  createSlice({
    name: scope,
    initialState,
    reducers: {
      resetState: resetState(initialState),
      ...reducers
    },
    extraReducers: {
      [asyncThunk.pending]: (state, action) => {
        state.error = null
        state.status = STATUS.PENDING

        if (action.meta.arg?.isDirty !== false) {
          state.data = []
        }
      },
      [asyncThunk.fulfilled]: (state, action) => {
        const {payload: {data = []} = {}} = action

        if (action.meta.arg?.isDirty === false) {
          state.data.push(...data)
        } else {
          state.data = data
        }

        state.hasNextPage = Boolean(action.payload?.hasNextPage)
        state.status = STATUS.SUCCESS
        state.total = action.payload?.total ?? 0
      },
      [asyncThunk.rejected]: (state, action) => {
        state.error = action.payload?.error

        if (action.meta.aborted !== true) {
          state.status = STATUS.FAILURE
        }
      },
      ...extraReducers
    }
  })
/* eslint-enable */

export {
  // eslint-disable-next-line import/prefer-default-export
  createAsyncResultsSlice
}
