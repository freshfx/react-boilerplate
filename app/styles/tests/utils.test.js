import {divide, addAlpha, createOpacityAnimation} from '../utils'

const two = 2

describe('styles utils', () => {
  it('should divide value with unit', () => {
    const fixture = '12px'
    const expectedResult = '6px'
    expect(divide(fixture, two)).toEqual(expectedResult)
  })

  it('should divide value without unit', () => {
    const fixture = '12'
    const expectedResult = '6'
    expect(divide(fixture, two)).toEqual(expectedResult)
  })

  it('should add alpha', () => {
    const white = '#fff'
    const alpha = 0.5
    expect(addAlpha(white, alpha)).toEqual('rgba(255,255,255,0.5)')
  })

  describe('createOpacityAnimation', () => {
    it('should create an animation with default params', () => {
      expect(createOpacityAnimation()).toMatchSnapshot()
    })

    it('should create an animation with the passed params', () => {
      expect(createOpacityAnimation(1, 0)).toMatchSnapshot()
    })
  })
})
