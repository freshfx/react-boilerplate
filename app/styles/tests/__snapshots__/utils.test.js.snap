// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styles utils createOpacityAnimation should create an animation with default params 1`] = `
Keyframes {
  "id": "sc-keyframes-bBzLxR",
  "inject": [Function],
  "name": "bBzLxR",
  "rules": Array [
    "@-webkit-keyframes bBzLxR{from{opacity:0;}to{opacity:1;}}",
    "@keyframes bBzLxR{from{opacity:0;}to{opacity:1;}}",
  ],
  "toString": [Function],
}
`;

exports[`styles utils createOpacityAnimation should create an animation with the passed params 1`] = `
Keyframes {
  "id": "sc-keyframes-jgOcXV",
  "inject": [Function],
  "name": "jgOcXV",
  "rules": Array [
    "@-webkit-keyframes jgOcXV{from{opacity:1;}to{opacity:0;}}",
    "@keyframes jgOcXV{from{opacity:1;}to{opacity:0;}}",
  ],
  "toString": [Function],
}
`;
